packages/devtools-core/src/utils/communication.ts

```diff
@@ -0,0 +1,41 @@
+import { Dispatch } from "@prodo/core/lib/types";
+import { state } from "../model";
+import { Action, AppMessage, DevMessage } from "../types";
+
+export const recordState = (newState: any) => {
+  state.app.state = newState;
+};
+
+export const recordAction = (action: Action) => {
+  state.app.actionLog.push(action);
+  if (action.rerender && Object.keys(action.rerender).length > 0) {
+    Object.keys(action.rerender).forEach(rerender => {
+      const renderLogEntry = {
+        componentId: rerender,
+        actionName: action.actionName,
+      };
+      state.app.renderLog.push(renderLogEntry);
+    });
+  }
+};
+
+export const eventListener = (dispatch: Dispatch) => (event: MessageEvent) => {
+  if (event.data.destination === "devtools") {
+    const message: DevMessage = event.data;
+    if (message.type === "state") {
+      dispatch(recordState)(message.contents.state);
+    } else if (message.type === "completedEvent") {
+      dispatch(recordAction)(message.contents.event);
+      if (message.contents.event.nextUniverse) {
+        dispatch(recordState)(message.contents.event.nextUniverse.state);
+      }
+    }
+  }
+};
+
+export const sendMessage = (
+  message: Partial<AppMessage> & Pick<AppMessage, "type">,
+) => {
+  const appMessage: AppMessage = { destination: "app", ...message };
+  window.postMessage(appMessage, "*");
```

Comment:
Where does this send the message too? I had no idea you could do this.

Reply:
This posts it to the general browser window, where it is then listened to by the devtools-plugin. That'll then trigger an action on the user app. If you're asking about the "*", it indicates no preference for the origin of the target window. (It is less secure though, and we should be able to get a more specific target - I can add that if that was the question)